todos-app/backend/user_service/entrypoint.sh

```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+set -e
+
+# Remove a potentially pre-existing server.pid for Rails.
+rm -f /app/tmp/pids/server.pid
+
+# Wait until DB is up and ready
+until mysql -h"${DB_HOST}" -u"${DB_USER}" -p"${DB_PASSWORD}" -e 'SELECT 1'
+do
+  echo "DB not ready!"
+  sleep 1
+done
```

Comment:
I have a question about this part, is it better to wait DB container like this than to set `depends_on` in `docker-compose.yml` ? 👀

Reply:
From what I know, `depends_on` only dictates the order of containers being created, but it doesn't really do anything to ensure the dependencies to be ready. So, while the DB container is created before the backend app, it takes a few seconds for it to be ready to accept connections. The backend container is created almost instantenously after DB is created, and it will try to connect to the unready DB and fails.

Another way to do it is by using `restart_policy` in docker-compose file like this;
```yaml
task-service:
    build: ./backend/task_service
    image: task-service
    container_name: task-service
    deploy:
      restart_policy:
        condition: on-failure
        delay: 3s
    ports:
      - 3001:3000
```
This will force the container to restart every 3 seconds after it fails, so eventually it will connect to DB.

I personally prefer having the retry logic on the container level rather than on docker-compose file so each container image does not depend on the docker-compose functionality 🙃

Reply:
Thank you for the detailed explanation!

> I personally prefer having the retry logic on the container level rather than on docker-compose file so each container image does not depend on the docker-compose functionality

I see, I've had trouble with the startup order before so this was helpful 🙌